### ISSUE TITLE
dbEvent: double cancel causes hang

### ISSUE THREAD

**Issue by mdavidsaver** [epics-base/epics-base#423]
A logic hole in dbEvent.c can combine with a race between concurrent `db_cancel_event()` and an event callback to cause `db_cancel_event()` to hang.  The following assertion will occasionally fail if a thread calls `db_cancel_event()` on two different subscriptions.  Not a double free, two separate free's of subscriptions which share the same dbEventCtx.

While I am not sure I can completely explain the sequence of events, I think this issue arises from an attempt to handle self-cancel and concurrent cancel with the same logic.  One fix I will explore is to move self-cancel detection from the common `struct event_user` to the individual `struct evSubscrip`.

```diff
diff --git a/modules/database/src/ioc/db/dbEvent.c b/modules/database/src/ioc/db/dbEvent.c
index 6dab45104..7cbbda91b 100644
--- a/modules/database/src/ioc/db/dbEvent.c
+++ b/modules/database/src/ioc/db/dbEvent.c
@@ -620,6 +620,7 @@ void db_cancel_event (dbEventSubscription event)
     assert ( pevent->npend == 0u );
 
     if ( pevent->ev_que->evUser->taskid == epicsThreadGetIdSelf() ) {
+        assert(!pevent->ev_que->evUser->pSuicideEvent);
         pevent->ev_que->evUser->pSuicideEvent = pevent;
     }
     else {
```

```
A call to 'assert(!pevent->ev_que->evUser->pSuicideEvent)'
    by thread 'qsrvSingle' failed in ../db/dbEvent.c line 623.
Dumping a stack trace of thread 'qsrvSingle':
[    0x7f5dcc2a352a]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libCom.so.3.22.1(epicsBackTrace+0x20)
[    0x7f5dcc2a33ac]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libCom.so.3.22.1(epicsStackTrace+0x6f)
[    0x7f5dcc299d21]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libCom.so.3.22.1(epicsAssert+0x5e)
[    0x7f5dcc31dc6b]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../../base-git/lib/linux-x86_64-debug/libdbCore.so.3.22.1(db_cancel_event+0x1fb)
[    0x7f5dcc6882a1]: /home/dev/epics/pvxs/test/O.linux-x86_64-debug/../../lib/linux-x86_64-debug/libpvxsIoc.so.1.2(_ZZN4pvxs3ioc12Subscription9subscribeEPvRKNS0_7ChannelEPFvS2_P9dbChanneliP12db_field_logES2_jENUlS2_E_clES2_+0x2b)
```

**Comment by mdavidsaver** [comment]
This situation came to light while @simon-ess is working on re-writing testpvalink with PVXS.  I can't see anything specific to PVXS, and both versions of QSRV use dbEvent in the same way (with a pair of evSubscript).  So it may turn out that some of the "false positive" failures of testpvalink were not entirely false...

**Comment by mdavidsaver** [comment]
My attempted fix in #424 has an issue with concurrent cancel.  Removing my big block comment from the `sync` branch in `db_cancel_event()` leaves:

```c
        curSeq = evUser->pflush_seq;
        do {
            epicsMutexUnlock( evUser->lock );
            epicsEventMustWait(evUser->pflush_sem);
            epicsEventTrigger(evUser->pflush_sem);
            epicsMutexMustLock ( evUser->lock );
        } while(curSeq == evUser->pflush_seq);
```

Which does not check if the worker has gone idle already.  Worse, the attempted re-trigger will start a tight loop.
